src/rust/engine/process_execution/src/docker.rs

```diff
@@ -295,222 +352,331 @@ impl super::CommandRunner for CommandRunner {
 
 #[async_trait]
 impl CapturedWorkdir for CommandRunner {
-  type WorkdirToken = (PathBuf, PathBuf);
+  type WorkdirToken = String;
 
   async fn run_in_workdir<'s, 'c, 'w, 'r>(
     &'s self,
     context: &'c Context,
-    workdir_path: &'w Path,
-    (immutable_inputs_workdir, named_caches_workdir): Self::WorkdirToken,
+    _workdir_path: &'w Path,
+    sandbox_path_in_container: Self::WorkdirToken,
     req: Process,
     _exclusive_spawn: bool,
   ) -> Result<BoxStream<'r, Result<ChildOutput, String>>, String> {
-    let docker = self.docker().await?;
+    let docker = self.docker.get().await?;
 
     let env = req
       .env
       .iter()
       .map(|(key, value)| format!("{}={}", key, value))
       .collect::<Vec<_>>();
 
-    let workdir_path_as_string = workdir_path
-      .to_path_buf()
+    let working_dir = req
+      .working_directory
+      .map(|relpath| Path::new(&sandbox_path_in_container).join(&relpath))
+      .unwrap_or_else(|| Path::new(&sandbox_path_in_container).to_path_buf())
       .into_os_string()
       .into_string()
       .map_err(|s| {
         format!(
-          "Unable to convert workdir_path due to non UTF-8 characters: {:?}",
+          "Unable to convert working directory due to non UTF-8 characters: {:?}",
           s
         )
       })?;
 
-    let immutable_inputs_workdir_as_string = immutable_inputs_workdir
+    let image = req
+      .docker_image
+      .ok_or("docker_image not set on the Process, but the Docker CommandRunner was used.")?;
+
+    // Obtain ID of the base container in which to run the execution for this process.
+    let container_id = self
+      .container_cache
+      .container_id_for_image(&image, &context.build_id)
+      .await?;
+
+    // DOCKER-TODO: Set creation options so we can set platform.
+
+    let config = bollard::exec::CreateExecOptions {
+      env: Some(env),
+      cmd: Some(req.argv),
+      working_dir: Some(working_dir),
+      attach_stdout: Some(true),
+      attach_stderr: Some(true),
+      ..bollard::exec::CreateExecOptions::default()
+    };
+
+    log::trace!("creating execution with config: {:?}", &config);
+
+    let exec = docker
+      .create_exec::<String>(&container_id, config)
+      .await
+      .map_err(|err| format!("Failed to create Docker execution in container: {:?}", err))?;
+
+    log::trace!("created execution {}", &exec.id);
+
+    let exec_result = docker
+      .start_exec(&exec.id, None)
+      .await
+      .map_err(|err| format!("Failed to start Docker execution `{}`: {:?}", &exec.id, err))?;
+    let mut output_stream = if let StartExecResults::Attached { output, .. } = exec_result {
+      output.boxed()
+    } else {
+      panic!(
+        "Unexpected value returned from start_exec: {:?}",
+        exec_result
+      );
+    };
+
+    log::trace!("started execution {}", &exec.id);
+
+    let exec_id = exec.id.to_owned();
+    let docker = docker.clone();
```

Comment:
Because `docker` here is `&Docker` and needs to be cloned to satisfy the borrow checker by obtaining an owned copy. Maybe it would be better to use `to_owned()` instead?

Reply:
Switch to `to_owned` to make it clearer.